### PR TITLE
Redux integration

### PR DESCRIPTION
**Pull request by yched**
As per the discussion in https://github.com/thereactivestack/kickstart-hugeapp/issues/27, here's a proposal for optional redux support :

Usage : 
Behavior is triggered by a 'createReduxStore' callback in clientOptions / serverOptions.
A 'wrapper' is also required, and it should be the react-redux `<Provider>` (or some custom component wrapping it)

```js
import { Provider } from 'react-redux'
import routes from './myAppRoutes'
import createStore from './myAppStore'

Meteor.startup(() => {
  const clientOptions = {
    wrapper: Provider, // Or some custom Component wrapping the redux Provider
    createReduxStore: (initialState, history) => {
      const store = createStore(initialState);
      // If using redux-simple-router, this is where you can bind the history with syncReduxAndRouter
      syncReduxAndRouter(history, store);
      return store;
    }
  });
  // Use the same 'wrapper' and 'createReduxStore' in serverOptions, or adjust to your needs.
  const serverOptions = clientOptions;
  ReactRouterSSR.Run(routes, clientOptions, serverOptions);
});
```

In detail, if options.createReduxStore callback is set :

On the server :
- After route is resolved and before rendering
  - create the store, passing a history to let the callback bind with redux-simple-router if it needs to
  - allow data fetching : let the components for the route fire actions (possibly async / Promise-based) to populate the store, and wait for them to be done.
    (see bottom of http://rackt.org/redux/docs/recipes/ServerRendering.html - implementation is inspired from https://github.com/erikras/react-redux-universal-hot-example)
  - pass the store as a prop to the wrapper component (the redux `<Provider>`)
- After rendering,
  - grab the resulting store state and pass it in the HTML (using meteorhacks:inject-data)

On the client :
- Before rendering 
  - create the store with the initial state populated from the server (if any)
  - pass the store as a prop to the wrapper component (the redux `<Provider>`)

### CODE REVIEW

File: lib/server.jsx [hunk 1]
```diff
@@ -183,14 +183,42 @@ function generateSSRData(serverOptions, context, req, res, renderProps) {
...
 
+      // Do the rendering.
       html = ReactDOMServer.renderToString(app);
 
+      // If using redux, pass the resulting redux state to the client so that it
+      // can hydrate from there.
+      if (reduxStore) {
+        // @todo JSON.parse(JSON.stringify()) is used to reduce possible immutables down to
+        // pure Javascript. There is probably a smarter way.
+        res.pushData('redux-initial-state', JSON.parse(JSON.stringify(reduxStore.getState())));
```
Comment (yched) [inline lib/server.jsx]: Still need to investigate.

Ideally we would just do `res.pushData('redux-initial-state', reduxStore.getState())`

But on my sample app the store contains ImmutablsJS data, and they don't seem to get along well with the EJSON.stringify() called by meteorhacks/inject-data.

```
(STDERR) iterable.length has been deprecated, use iterable.size or iterable.count(). This warning will become a silent error in a future version. Error
     at src_Map__Map.Object.defineProperty.get (webpack:///.meteor/local/webpack-npm/~/immutable/dist/immutable.js:4633:14)
     at looksLikeArray (packages/underscore/underscore.js:94:1)
     at Function._.each._.forEach (packages/underscore/underscore.js:106:1)
     at Object.EJSON.clone (packages/ejson/ejson.js:500:1)
     at packages/ejson/ejson.js:501:1
     at Function._.each._.forEach (packages/underscore/underscore.js:113:1)
     at Object.EJSON.clone (packages/ejson/ejson.js:500:1)
     at packages/ejson/ejson.js:501:1
     at Function._.each._.forEach (packages/underscore/underscore.js:113:1)
     at Object.EJSON.clone (packages/ejson/ejson.js:500:1)
```

**Comment by genyded** [comment]
If we can get this integrated in an official build - that would be cool. The Immutable JS stuff can be dealt with in other ways and not everyone is using it.

**Comment by yched** [comment]
@genyded : agreed, but IMO we're fine, the current code works whether there are immuables or not.

If there are, they are downcasted to raw JSON data before being sent to the client (this "just works" with JSON.stringify(), but not with Meteor's EJSON.stringify(), because of the way it's coded). The client will need to implement some logic to re-upcast to Immutable structures on rehydrate - AFAIK that is the usual behavior for immutables and ssr, and that is usually done in the reducers (redux-immutable's createReducer() has a 'converter' param just for that case), as part of the regular `createStore(initialDataAsRawJSON)`, so this is not our concern here.

In short : we're not doing extra work for Immutables, we just do this one `res.pushData()` line in a way that does not break them.

**Comment by eXon** [comment]
Awesome work. I'll try to review it ASAP. I haven't tried Redux outside of vanilla Node.js so I'm thrilled to start using it with Meteor :)

**Comment by yched** [comment]
Swapped the order of arguments for the createReduxStore() callback to (initialState, history), that should be slightly more intuitive on the consuming side :
- initialState usually comes first in redux createStore() functions,
- and not everyone will care about the history param

I adjusted the "Usage" instructions in the PR description above accordingly.

**Comment by eXon** [comment]
Awesome job! I'll merge this but first you need to document this on the readme :)

**Comment by yched** [comment]
Added some doc - might need some proof-reading :-)

**Comment by dbismut** [comment]
Hey @yched, I'm very new to all this, so apologies for any mistakes below.

I tried using your branch as a local package to my app which uses the `hugeapp` bootstrap. There seems to be an issue with `InjectData` and you might be missing the `https://github.com/meteorhacks/inject-data/` in `package.js`?

Also, if the `initialData` is null, I got an err `Unexpected token u` in the console, which seems to be relating to `JSON.parsing` a null object. Changing ([L25 of client.jsx](https://github.com/yched/meteor-react-router-ssr/blob/redux-integration/lib/client.jsx#L25) to:

```
clientOptions.createReduxStore(null, history);
```
... does the trick but I guess you should check `data` is defined before parsing it.

**Comment by yched** [comment]
@dbismut : Thanks for the feedback !

> There seems to be an issue with InjectData and you might be missing the meteorhacks:inject-data package in package.js?
Hm, meteor-react-router-ssr depends on meteorhacks:fast-render, which itself depends on meteorhacks:inject-data, so that should be fine ?
I used meteorhacks:inject-data's res.pushData() (server) and InjectData.getData() (client) because they were "already there" :-)

> Also, if the initialData is null, I got an err `Unexpected token u` in the console, which seems to be related to JSON.parsing a null object

Good call, should be fixed by the commit above.

**Comment by dbismut** [comment]
Well, I don't know, the console threw an error with InjectData not being defined. My understanding is that a meteor package dependency is only exposed to the package that relies on it and not exposed to the rest of the app unless `api.imply` is used... Since fast-render doesn't imply inject-data I'm not sure why InjectData would be available in react-router-ssr. 

But hey I might be wrong :)

**Comment by yched** [comment]
@dbismut : Yup, I guess you're right. TBH I'm fairly new to the Meteor Packages API myself ;-) It seems a package can't implicitly rely on its dependencies' dependencies, and instead needs to explicitly "api.use()" all the packages it direclty uses - makes sense.

The code in this PR actually comes a sample app of mine that uses a copy of the meteor-react-router-ssr code but doesn't use the *package* itself (because the package brings all the meteor wrapper packages for react, router, helmet, etc..., and for now I'd rather try to use the "regular" npm versions directly through webpack:webpack). So, while the code does work, I probably missed the "api.use()" implications in meteor-react-router-ssr's package.json :-/

Should be fixed in the commit above.

In the same spirit, I also added a weak dependency on the 'promise' package, since it's only used for the (optional) "pre-render fetchData()" feature of the (optional) redux integration. I just do a runtime check for `if (!Package.promise) {console.error(...)}` if that feature is used. A full, hard dependency on 'promise' wouldn't be that much different since that package is installed pretty much everywhere anyway (it's a dependency of the 'ecmascript' package), but a weak dependency seems more accurate. I'll let @eXon chime in if he feels different.

File: README.md [hunk 2]
```diff
@@ -98,3 +100,63 @@ if (Meteor.isClient) {
...
+Meteor.startup(() => {
+  const clientOptions = {
+    wrapper: Provider,
+    createReduxStore: (initialState, history) => {
+      const store = createStore(initialState);
+      // If using redux-simple-router, this is where you can bind the history with syncReduxAndRouter
+      syncReduxAndRouter(history, store);
+      return store;
+    }
+  });
```
Comment (eXon) [inline README.md]: Their is a parenthesis here that shouldn't be

File: lib/server.jsx [hunk 3]
```diff
@@ -183,14 +183,43 @@ function generateSSRData(serverOptions, context, req, res, renderProps) {
         ...serverOptions.props
       };
 
-      let app = <RoutingContext {...renderProps} />;
+      // If using redux, create the store.
+      let reduxStore;
+      if (typeof serverOptions.createReduxStore !== 'undefined') {
+        // Create a history and set the current path, in case the callback wants
+        // to bind it to the store using redux-simple-router's syncReduxAndRouter().
+        const reduxHistory = history.useQueries(history.createMemoryHistory)();
```
Comment (eXon) [inline lib/server.jsx]: history is not defined anywhere here

**Comment by eXon** [comment]
You did an awesome job and it is really close to be merge :) Once you fix the `history` undefined, it looks good!

**Comment by yched** [comment]
Oops, indeed. Fixed.

**Comment by dbismut** [comment]
Thank you! Now I can try redux. I learned "modern" JavaScript with meteor and have been reading a lot from @eXon in particular on how it doesn't necessarily makes sense with meteor, but if you're not using it, how do you decouple the logic from the components is not clear to me. Anyway thanks again :)

**Comment by dbismut** [comment]
I just have a question, not sure this is the right place, but is it possible to have a different actions and reducers (and store?) depending on your webpack chunk? With the same logic that I don't want the admin chunk loaded when not needed, I would not want all admin-related actions to be loaded when not needed...

**Comment by eXon** [comment]
Dan (the creator of Redux) has this answer on Stack Overflow about it. It might have some side-effects AFAIK.

http://stackoverflow.com/questions/32968016/how-to-dynamically-load-reducers-for-code-splitting-in-a-redux-application

**Comment by dbismut** [comment]
Thanks!!

**Comment by genyded** [comment]
So where are we with merging this?

**Comment by eXon** [comment]
After doing some review and testing, this seems to work pretty nicely. Thanks for your contribution!

**Comment by yched** [comment]
Yay :-)

**Comment by dbismut** [comment]
Hey again @yched! So I'm now trying to integrate `redux-simple-router` and therefore `syncReduxAndRouter`. However, reading [this](https://github.com/rackt/redux-simple-router/issues/31), it looks like `redux-simple-router` is not made for server-side rendering, and indeed, it throws `error while server-rendering ReferenceError: window is not defined` in the console.

From the [Example with Redux in the ReadMe](https://github.com/thereactivestack/meteor-react-router-ssr#example-with-redux), it says:

```javascript
Meteor.startup(() => {
  const clientOptions = {
    wrapper: Provider,
    createReduxStore: (initialState, history) => {
      const store = createStore(initialState);
      // If using redux-simple-router, this is where you can bind the history with syncReduxAndRouter
      syncReduxAndRouter(history, store);
      return store;
    }
  };
  // Use the same 'wrapper' and 'createReduxStore' in serverOptions, or adjust to your needs.
  const serverOptions = clientOptions;

  ReactRouterSSR.Run(routes, clientOptions, serverOptions);
});
```

I guess `serverOptions = clientOptions` is the culprit here, so you should probably add `Meteor.isClient` to run the redux router sync only on the client...? This is fixing it for me, but I'm only starting redux so I haven't tested much of anything yet.

```javascript
if (Meteor.isClient) {
      syncReduxAndRouter(history, store);
}
```

**Comment by yched** [comment]
@dbismut : hm, I don't seem to have that issue on my sample app.
I remember having similar issues when using a history object created with the "usual" createHistory(), which assumes a browser, but the server-side version of ReactRouterSSR now uses createMemoryHistory(), which works fine.

Can you provide a more detailed callstack for the "window is not defined" error ?
(also, which versions of redux-simple-router / react-router / history do you use ?)

My take is that using syncReduxAndRouter() on the server-side as well is useful because it lets the "@@router/INIT_PATH" action get fired once routing is resolved, and correctly populates the store's 'routing' element with the expected info before your route component gets rendered, just like it would be on the client.

**Comment by dbismut** [comment]
@yched: I must be wrong, again, I'm very new to all this.
Anyway, I'm using `"redux-simple-router": "^1.0.2"` within the `webpack.packages.json` file and `react-router, history` are the npm packages used by `meteor-react-router` (which I believed are the latest versions - 1.0.3 and 1.17.0 respectively).

Anyway, here is the complete call stack:

```
error while server-rendering ReferenceError: window is not defined
     at webpack:///Users/david/repo/tuly-react/.meteor/local/webpack-npm/~/redux-form/~/deep-equal/index.js:26:1
     at Object.dispatch (webpack:///Users/david/repo/tuly-react/.meteor/local/webpack-npm/~/redux-form/~/deep-equal/index.js:26:1)
     at webpack:///Users/david/repo/tuly-react/modules/producer/client/components/sidebar.jsx:43:20
     at node_modules/history/lib/useQueries.js:77:1
     at Object.listen (node_modules/history/lib/createHistory.js:98:1)
     at Object.listen (node_modules/history/lib/useQueries.js:76:1)
     at c (webpack:///Users/david/repo/tuly-react/modules/producer/client/components/sidebar.jsx:150:44)
     at Object.m.createReduxStore (webpack:///server.js:9136:370)
     at lib/server.jsx:206:36
     at [object Object]._.extend.withValue (packages/meteor/dynamics_nodejs.js:56:1)
```

And here is my implementation:

```javascript
// store.js
const middlewares = [
  thunkMiddleware,
  loggerMiddleware
];

const finalCreateStore = compose(
  applyMiddleware(...middlewares),
)(createStore);

export default function configureStore(history, initialState) {
  const store = finalCreateStore(rootReducer, initialState);
  return store;
}

// routes.js
import configureStore from './store';

const clientOptions = {
  wrapper: Provider,
  createReduxStore: (initialState, history) => {
    const store = configureStore(history, initialState);
    syncReduxAndRouter(history, store);
    return store;
  }
};

const serverOptions = {
  ...clientOptions,
  preRender: function (req, res) {
    // this is for react-material-ui to work with SSR 
    GLOBAL.navigator = {
      userAgent: req.headers['user-agent']
    };
  }
};
```

**Comment by dbismut** [comment]
Forgot to mention, this obviously happens when you run in production mode, otherwise SSR is disabled so the error isn't thrown.

**Comment by genyded** [comment]
What are you actually rendering? It looks like you may have redux-form in the mix, and if so that may be the culprit... don't think forms can render server-side. Can you render a simple H1 or something as a test? If that works, it's likely something in the specific render.

**Comment by dbismut** [comment]
Hey @genyded, well I'm rendering several different components, I think most of them are pure. In any case, they don't make use of `jQuery` or `windows` or anything that would throw an error server-side... However, without `syncReduxAndRouter` in `the clientOptions` for `react-router-ssr` everything works fine with SSR, so I don't see why `redux-form` would not work (although I understand it shows in the call stack ^^).

I have two chunks and in one chunk, I'm rendering just some divs.



 It stills throws an error in the console, and the response from the server is `<div id="react-app">undefined</div>`, then the client overrides and renders properly the capture above.

**Comment by yched** [comment]
Weird - but yeah, it would be good to try without redux-form, with an app containing only a very silly component ?

**Comment by dbismut** [comment]
Sure give me day, I'll try.

**Comment by genyded** [comment]
Just thought I would chime in and say I am using this with redux-simple-router (with any client wrapper) and it is working fine for me too. I am using devtools and all the routes/history show up there as expected as well. I am using browserHistory instead of the default which I believe is hash...

`import { browserHistory as history } from 'react-router'`

...but I have everything else pretty much per the readme example including the syncReduxAndRouter, but I am passing in the initial router state there as well...

`syncReduxAndRouter(history, store, state => state.router)`

**Comment by dbismut** [comment]
@genyded well apparently `react-router-ssr` uses `ReactRouter.history.useQueries(ReactRouter.history.createMemoryHistory)();` to create the history (which is not `createHashHistory` but is not `createBrowserHistory` either which is the recommended option: @yched any reason why you chose `createMemoryHistory`?).

In any case, [as in the docs](https://github.com/rackt/redux-simple-router#syncreduxandrouterhistory-store-selectrouterstate), `syncReduxAndRouter` does already the mapping to `state.routing` provided you install the reducer under the name "routing" (which is not your case so I understand why you might want to pass `state => state.router`.

**Comment by yched** [comment]
@genyded : awesome, thanks for the feedback !

Note that the createHistory() function exported by the 'history' package is actually a wrapper for createBrowserHistory()

@dbismut : I went with MemoryHistory on the server-side because AFAICT that's the only one that doesn't assume a browser ? createHashHistory and createBrowserHistory both assume there's a `window` - see https://github.com/rackt/history/blob/master/docs/GettingStarted.md

In short: 
- by default ReactRouterSSR uses a BrowserHistory on the client-side, if no explicit history object is passed in clientOptions - that was the pre-existing behaviour, the "redux integration" PR didn't touch that part.
- on the server-side, it uses a MemoryHistory (at the moment, only needed for redux support if a createReduxStore option is present in the serverOptions). I couldn't think of a use case for allowing an explicit history object to be specified in serverOptions like on the client-side, so I didn't add that to avoid needless confusion, but that could be something to reconsider if needed :-)

**Comment by dbismut** [comment]
Ok very clear, thanks! Indeed I only checked server-side code, that was stupid of me sorry 😶
I'll investigate on my issue then.

**Comment by minfawang** [comment]
@yched Could you please explain a little more of the injection of initialState. The only relevant part I see is:
```javascript
      InjectData.getData('redux-initial-state', data => {
        const initialState = data ? JSON.parse(data) : undefined;
        reduxStore = clientOptions.createReduxStore(initialState, history);
```
but it still looks like magic to me. Is there a way we could change the initial state passed in the createStore?

**Comment by yched** [comment]
@minfawang look at server.js in the PR. The initial state passed to the client is the one read from the store after routing and rendering was done on the server-side.

More importantly, that includes results of the actions dispatched by the optional "fetchData()" static method on the route's component to fetch external data if needed before rendering (see the README). Aside from that "fetchData()" step.

Not sure if there is a case to further modify that state before sending it to the client. In redux, the state is only supposed to be changed by firing actions, and that's what fetchData() is here for.

**Comment by minfawang** [comment]
@yched Thanks for your prompt response!! I just realized the existence of `fetchData`. That's very handy!  :smile:

**Comment by dbismut** [comment]
Ok back on https://github.com/thereactivestack/meteor-react-router-ssr/pull/24#issuecomment-169299351. Wow. This took way more time that I thought to realize I'm super stupid. The culprit is the `loggerMiddleware`. It prints to the browser console, so needs to be deactivated in production for SSR. Obviously. Sorry for this :tired_face:

**Comment by yched** [comment]
@dbismut : :-)
Weird that the callstack didn't seem to provide actual hints that your loggerMiddleware was at fault...

Note : [redux-logger](https://github.com/fcomb/redux-logger) seems to work pretty fine on both client and server side, and produces nicely formatted messages in both places. I used it that way :

```js
// myLoggerMiddleware.js
import createLogger from 'redux-logger'
const options = {
    logger: console,
    duration: true,
    collapsed: true,
    colors: Meteor.isClient
}
export default createLogger(options);

**Comment by dbismut** [comment]
Hmm will have a look, thanks!
